php/class-customize-snapshot-manager.php

```diff
@@ -934,4 +943,73 @@ public function get_front_uuid_param() {
 	public function get_customize_uuid_param() {
 		return constant( get_class( $this->post_type ) . '::CUSTOMIZE_UUID_PARAM_NAME' );
 	}
+
+	/**
+	 * Save the preview url query vars in changeset meta.
+	 *
+	 * @param int $post_id Post id.
+	 */
+	public function save_customize_preview_url_query_vars( $post_id ) {
+		if ( ! isset( $_POST['customize_preview_url_query_vars'] ) ) {
+			return;
+		}
+
+		$original_query_vars = (array) json_decode( wp_unslash( $_POST['customize_preview_url_query_vars'] ) );
+
+		if ( empty( $original_query_vars ) ) {
+			return;
+		}
+
+		$allowed_panel_section_control_params = array(
+			'autofocus[panel]',
+			'autofocus[section]',
+			'autofocus[control]',
+		);
+
+		$allowed_query_params = array_merge( $allowed_panel_section_control_params, array(
+			'device',
+			'scroll',
+		) );
+
+		$allowed_devices = array(
+			'mobile',
+			'desktop',
+			'tablet',
+		);
+
+		$preview_url_query_vars = array();
+		$params = wp_array_slice_assoc( $original_query_vars, $allowed_query_params );
+
+		if ( ! empty( $params ) ) {
+			foreach ( $params as $param => $value ) {
+				$is_valid_var = (
+					(
+						in_array( $param, $allowed_panel_section_control_params )
+						&&
+						preg_match( '/[a-z|\[|\]|_|-|0-9]+/', $value )
```

Comment:
Isn't the purpose to validate that the `$param` is a valid ID string? If you want to ensure that _only_ the valid characters are included then you should do:

```php
preg_match( '/^[a-z|\[|\]|_|-|0-9]+$/', $value )
```

Adding the `^` and `$` anchors, as otherwise a user could supply an ID like: `!@#$%^&*()title_tagline⁄€‹›ﬁﬂ‡°·‚`